references/classification/train_quantization.py

```diff
@@ -84,13 +109,20 @@ def main(args):
             ds, batch_size=args.batch_size, shuffle=False, num_workers=args.workers, pin_memory=True
         )
         model.eval()
-        model.fuse_model(is_qat=False)
-        model.qconfig = torch.ao.quantization.get_default_qconfig(args.backend)
-        torch.ao.quantization.prepare(model, inplace=True)
+        if use_fx_graph_mode_quantization:
+            qconfig_dict = torch.ao.quantization.get_default_qconfig_dict(args.backend)
+            model = torch.ao.quantization.quantize_fx.prepare_fx(model, qconfig_dict)
+        else:
+            model.fuse_model(is_qat=False)
+            model.qconfig = torch.ao.quantization.get_default_qconfig(args.backend)
+            torch.ao.quantization.prepare(model, inplace=True)
         # Calibrate first
         print("Calibrating")
         evaluate(model, criterion, data_loader_calibration, device=device, print_freq=1)
-        torch.ao.quantization.convert(model, inplace=True)
+        if use_fx_graph_mode_quantization:
+            model = torch.ao.quantization.quantize_fx.convert_fx(model)
+        else:
+            torch.ao.quantization.convert(model, inplace=True)
```

Comment:
Since the updated version of this PR, no longer touches the `torchvision.models.quantization` methods, it's unclear how one would use models trained with the FX-based approach. Aka:
- How do we construct quantized versions of the models?
- How do we load the weights?

From what I understand in the Post-Training-Quantization case, the user would just do the following:
```python
# 1. Construct pre-trained unquantized model
model = torchvision.models.resnet50(weights=ResNet50_Weights.IMAGENET1K_V1)
model.eval()

# 2. Quantize it 
qconfig_dict = torch.ao.quantization.get_default_qconfig_dict("qnnpack")
model = torch.ao.quantization.quantize_fx.prepare_fx(model, qconfig_dict)
model = torch.ao.quantization.quantize_fx.convert_fx(model)

# 3. Use it as normal
output = model(input)
```

Because step 2 is identical to all models, it can be a utility and thus there is no need for a custom quantized model builder (aka `torchvision.models.quantized.resnet50()`). Correct?

For QAT, I believe we need a quantized model builder because one needs to do:
```python
# 1. Construct an non-initialized model 
model = torchvision.models.resnet50(weights=None)
model.eval()

# 2. Convert it to FX
model = torch.ao.quantization.quantize_fx.convert_fx(model)

# 3. Load the QAT weights
weights = ResNet50_Weights.IMAGENET1K_QNNPACK_V1
model.load_state_dict(weights.get_state_dict(progress=True))

# 4. Use it as normal
output = model(input)
```

For QAT, the loading of the weights (step 3) must be performed after FX conversion (step 2) which means that now a custom quantized model builder is necessary to make it convenient to the users. Right?

Reply:
Hi @datumbox, actually both post-training quantization and QAT code paths are the same for FX and eager modes. In both cases, the workflows are as follows (pseudocode):
```
# PTQ
model = prepare(model) # or prepare_fx
calibrate(model, data)
model = convert(model) # or convert_fx

# QAT
model = prepare_qat(model) # or prepare_qat_fx
train_one_epoch(model, data)
evaluate(model, data)
model = convert(model) # or convert_fx
```
The main difference between FX and eager modes is which model is used. In eager mode, the user needs to manually at `QuantStub`s and `DeQuantStub`s, so they must use the `torchvision.models.quantization` version. In FX mode, however, the user can simply quantize the model as is (assuming it is symbolically traceable). This is why I used the unmodified `torchvision.models` version for FX mode in this PR.

> Because step 2 is identical to all models, it can be a utility and thus there is no need for a custom quantized model builder (aka torchvision.models.quantized.resnet50()). Correct?

Correct. This is why we use the original `torchvision.models` for FX quantization. However, we still need the custom quantized model builders (`torchvision.models.quantization`) for eager mode quantization.

> For QAT, the loading of the weights (step 3) must be performed after FX conversion (step 2) which means that now a custom quantized model builder is necessary to make it convenient to the users. Right?

Actually if there is no training involved and the goal of the user is to simply load quantized weights then they can just use PTQ. For e.g., they would simply initialize the model with these [quantized weights](https://github.com/pytorch/vision/blob/main/torchvision/models/quantization/resnet.py#L178) instead.

For QAT, my understanding is the user would only load the weights in the beginning _before_ training (and `convert`) as part of loading from pretrained weights. For FX QAT, however, there is still no need for a custom quantized model builder (i.e. `torchvision.models.quantization`) because the user doesn't have to make any manual modifications to the models in order to quantize them.

Please let me know if there is anything unclear in my explanation or if I misunderstood any part of your questions.

Reply:
Thanks a lot for the reply.

The examples you provided are great. I think the training part is clear to me after reading the PR. What I would like to understand better is how PTQ and QAT weight are loaded in a model. I want to confirm that indeed there is no need to add new model builders at `torchvision.models.quantization` for the FX case and that I don't need to modify any of the existing model builders at `torchvision.models`.

So assume I'm using the original builders for `torchvision.models` without any modification. What steps if any do I have to take to load PTQ/QAT weights? Example:

```python
# 1. Construct an non-initialized model 
model = torchvision.models.resnet50(weights=None)
model.eval()

# 2a. Load PTQ weights
# TODO

# 2b. Load QAT weights
# TODO
```

Could you fill the TODOs for me?

Reply:
Hi @datumbox, I'm continuing the discussion in the main thread. My comment there should clarify when the weights are loaded in different modes (PTQ/QAT/test_only).

Reply:
No problem, let's continuing the discussion in the main thread. My comment there should clarify when the weights are loaded in different modes (PTQ/QAT/test_only).